File: 백준/Silver/20301. 반전 요세푸스/반전 요세푸스.py
import sys
from collections import deque

input = sys.stdin.readline

n, k, m = map(int, input().split())

queue = deque(range(1, n + 1))

isRight = True  # 초기 오른쪽 방향으로 인덱스 검사
ret = []  # 결과 넣어둘 곳
index = 0

while queue:
    # k번째가 되게 회전
    if isRight:
        # 제거된 것 기준 오른쪽이라 이건 그냥 k-1로 해도 됨
        queue.rotate(-((k - 1) % len(queue)))
    else:
        # 제거된 것 기준 왼쪽으로 k번재인데 제거된게 없어서 k로 접근해야 함
        queue.rotate(k % len(queue))

    ret.append(queue.popleft())

    if (len(ret)) % m == 0:
        # m명이 제거되었다면 방향 전환
        isRight = not isRight
print('\n'.join(map(str, ret)))

File: 백준/Silver/20301. 반전 요세푸스/반전 요세푸스.kt
package org.example

fun main() {
    val br = System.`in`.bufferedReader()
    val sb = StringBuilder()
    val (n, k, m) = br.readLine().split(" ").map { it.toInt() }
    val queue = ArrayDeque<Int>((1..n).toList())
    val ret = mutableListOf<Int>()

    var isRight = true
    while (queue.isNotEmpty()) {
        if (isRight) {
            // 정방향 진행
            // 앞의 k-1개 빼서 뒤로 넣기
            repeat((k - 1) % queue.size) {
                queue.addLast(queue.removeFirst())
            }
        } else {
            // 반대방향 진행
            // 제거된 것 기준 고려
            repeat(k % queue.size) {
                queue.addFirst(queue.removeLast())
            }
        }
        ret.add(queue.removeFirst())
        if (ret.size % m == 0) {
            // m명 제거되면 방향 전환
            isRight = !isRight
        }
    }
    for (r in ret) {
        sb.append("$r\n")
    }
    print(sb)

}